src/v/raft/replication_monitor.cc

```diff
@@ -0,0 +1,191 @@
+              }
+          });
+    });
+}
+
+std::ostream&
+operator<<(std::ostream& o, const replication_monitor::waiter& entry) {
+    fmt::print(o, "{{type: {}, append result: {}}}", entry.type, entry.result);
+    return o;
+}
+
+std::ostream&
+operator<<(std::ostream& o, const replication_monitor::wait_type& type) {
+    switch (type) {
+    case replication_monitor::commit:
+        fmt::print(o, "commit");
+        break;
+    case replication_monitor::majority_replication:
+        fmt::print(o, "majority_replication");
+        break;
+    }
+    return o;
+}
+
+ss::future<> replication_monitor::stop() {
+    auto f = _gate.close();
+    for (auto& [_, waiter] : _waiters) {
+        waiter.done.set_value(errc::shutting_down);
+    }
+    _committed_event_cv.broken();
+    _replicated_event_cv.broken();
+    return f;
+}
+
+ss::future<errc> replication_monitor::do_wait_until(
+  storage::append_result append, wait_type type) {
+    if (_gate.is_closed()) {
+        return ssx::now(errc::not_leader);
+    }
+    auto it = _waiters.emplace(append.base_offset, waiter(append, type));
+    return it->second.done.get_future();
+}
+
+ss::future<errc>
+replication_monitor::wait_until_committed(storage::append_result append) {
+    return do_wait_until(append, wait_type::commit);
+}
+
+ss::future<errc> replication_monitor::wait_until_majority_replicated(
+  storage::append_result append) {
+    return do_wait_until(append, wait_type::majority_replication);
+}
+
+ss::future<> replication_monitor::do_notify_replicated() {
+    while (!_gate.is_closed()) {
+        auto majority_replicated_offset = _raft->_majority_replicated_index;
+        auto it = _waiters.begin();
+        while (it != _waiters.end()
+               && it->first <= majority_replicated_offset) {
+            auto& entry = it->second;
+            auto replicated_term = _raft->get_term(entry.result.last_offset);
+            if (
+              entry.type == wait_type::majority_replication
+              // We need the offset is majority replicated in the same term
+              // it is appended for the replication to be successful.
+              // Truncation is detected based on committed offset updates.
+              && replicated_term == entry.result.last_term
+              && majority_replicated_offset >= entry.result.last_offset) {
+                entry.done.set_value(errc::success);
+                it = _waiters.erase(it);
+            } else {
+                ++it;
+            }
+        }
+        co_await _replicated_event_cv.wait();
+    }
+}
+
+ss::future<> replication_monitor::do_notify_committed() {
+    while (!_gate.is_closed()) {
+        auto committed_offset = _raft->committed_offset();
+        auto committed_offset_term = _raft->get_term(committed_offset);
+        auto it = _waiters.begin();
+        while (
+          it != _waiters.end()
+          && (it->first <= committed_offset || it->second.result.last_term < committed_offset_term)) {
```

Comment:
nit: it maybe worth a comment that in the loop we are guaranteed that committed offset is greater than or equal to append result base offset.